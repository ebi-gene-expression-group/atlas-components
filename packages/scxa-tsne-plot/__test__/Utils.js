import HighchartsSeriesGenerator from 'highcharts-series-generator'

const seriesNames = [`0`, `1`, `2`, `3`, `4`]
const maxPointsPerSeries = 1000

const gradientColourRanges = () => {
  return [
    {
      colour: `rgb(215, 255, 255)`,
      threshold: 0,
      stopPosition: 0
    },
    {
      colour: `rgb(128, 255, 255)`,
      threshold: 10,
      stopPosition: 20
    },
    {
      colour: `rgb(0, 85, 225)`,
      threshold: 100,
      stopPosition: 40
    },
    {
      colour: `rgb(0, 0, 115)`,
      threshold: 10000,
      stopPosition: 100
    }
  ]
}

const plotData = (series) => {
  const allPoints = series.reduce((acc, series) => acc.concat(series.data), [])

  return {
    series: series,
    min: Math.min(...allPoints.map((point) => point.expressionLevel)),
    max: Math.max(...allPoints.map((point) => point.expressionLevel)),
    unit: `TPM`
  }
}

const randomHighchartsSeries = () => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithNamesAndMaxPoints = (seriesNames, maxPointsPerSeries) => {
  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries)
}

const randomHighchartsSeriesWithSeed = () => {
  const seed = `A hair, Morty. I need one of your hairs. This isnâ€™t Game of Thrones.`

  return HighchartsSeriesGenerator.generate(seriesNames, maxPointsPerSeries, seed)
}

export { plotData, gradientColourRanges, randomHighchartsSeries, randomHighchartsSeriesWithSeed, randomHighchartsSeriesWithNamesAndMaxPoints}