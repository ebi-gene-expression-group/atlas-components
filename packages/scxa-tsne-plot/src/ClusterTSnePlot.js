import React from 'react'
import PropTypes from 'prop-types'
import Color from 'color'
import {find as _find, flatten as _flatten} from 'lodash'

import ScatterPlotLoader from './plotloader/PlotLoader'
import PlotSettingsDropdown from './PlotSettingsDropdown'

const _colourizeClusters = (highlightSeries) =>
  (series) => series.map((aSeries) => {
    // I canâ€™t think of a better way to reconcile series.name being a string and highlightSeries being an array of
    // numbers. For more flexibility we might think of having our series be identified by an arbitrary ID string
    if (!highlightSeries.length || highlightSeries.length === series.length || highlightSeries.map((hs) => `Cluster ${hs}`).includes(aSeries.name)) {
      return aSeries
    } else {
      return {
        name: aSeries.name,
        data: aSeries.data.map((point) => ({
          ...point,
          color: Color(`lightgrey`).alpha(0.65).rgb().toString()
        }))
      }
    }
  })

const tooltipHeader = (clusterType, series, point) => {
  const clusterName = clusterType === `clusters` ?
    `<b>Cluster name:</b> ${series.name}<br>` : ``
  return `<b>Cell ID:</b> ${point.name}<br>` + clusterName
}

const ClusterTSnePlot = (props) => {
  const {ks, perplexities, metadata, selectedPerplexity, onChangePerplexity, selectedColourBy, onChangeColourBy, clusterType} = props  // Select
  const {plotData, highlightClusters, height, tooltipContent} = props   // Chart
  const {loading, resourcesUrl, errorMessage, showControls} = props   // Overlay
  const opacity = 0.7

  const highchartsConfig = {
    plotOptions: {
      scatter: {
        marker: {
          symbol: `circle`
        }
      }
    },
    // Generated with http://tools.medialab.sciences-po.fr/iwanthue/
    colors: [
      `rgba(212, 137, 48, ${opacity})`,
      `rgba(71, 193, 152, ${opacity})`,
      `rgba(90, 147, 221, ${opacity})`,
      `rgba(194, 73, 97, ${opacity})`,
      `rgba(128, 177, 66, ${opacity})`,
      `rgba(208, 76, 134, ${opacity})`,
      `rgba(188, 176, 59, ${opacity})`,
      `rgba(132, 43, 102, ${opacity})`,
      `rgba(93, 188, 108, ${opacity})`,
      `rgba(82, 45, 128, ${opacity})`,
      `rgba(101, 133, 52, ${opacity})`,
      `rgba(169, 107, 212, ${opacity})`,
      `rgba(185, 140, 70, ${opacity})`,
      `rgba(82, 88, 180, ${opacity})`,
      `rgba(176, 73, 62, ${opacity})`,
      `rgba(101, 127, 233, ${opacity})`,
      `rgba(214, 126, 188, ${opacity})`,
      `rgba(196, 86, 178, ${opacity})`,
      `rgba(173, 131, 211, ${opacity})`,
      `rgba(193, 84, 47, ${opacity})`
    ],
    chart: {
      height: height
    },
    title: {
      text: `Clusters`
    },
    legend: {
      enabled: true,
      align: `center`,
      verticalAlign: `bottom`,
      layout: `horizontal`,
      itemMarginBottom: 20
    },
    tooltip: {
      style: {
        width: `200px`,
        overflow: `auto`,
        whiteSpace: `normal`
      },
      formatter: function (tooltip) {
        // Trick Highcharts into thinking the point is in the bottom half of the chart, so that the tooltip
        // is displayed below the point
        this.point.negative = true
        const text = `Loading metadata...`
        const header = tooltipHeader(clusterType, this.series, this.point)
        const addMetadataToTooltipText = async () => {
          try {
            const response = await tooltipContent(this.point.name)

            const content = response.map((metadata) => {
              return `<b>${metadata.displayName}:</b> ${metadata.value}`
            })

            tooltip.label.attr({
              text: header + content.join(`<br>`)
            })
          } catch (e) {
            tooltip.label.attr({
              text: header
            })
          }
        }

        addMetadataToTooltipText()

        return header + text
      }
    }
  }

  const perplexityOptions = perplexities.sort((a, b) => a - b).map((perplexity) => ({
    value: perplexity,
    label: perplexity
  }))

  const kOptions = ks.sort((a, b) => a - b).map((k) => ({
    value: k.toString(),
    label: `k = ${k}`,
    group: `clusters`
  }))

  const metadataOptions = metadata.map((metadata) => ({
    ...metadata,
    group: `metadata`
  }))

  const options = [
    {
      label: `Metadata`,
      options: metadataOptions,
    },
    {
      label: `Number of clusters`,
      options: kOptions,
    },
  ]

  const defaultValue = _find(
    _flatten(
      options.map((item) => (item.options))
    ),
    {value: selectedColourBy}
  )

  return [
    <div key={`perplexity-k-select`} className={`row`}>
      {showControls &&
      <div className={`small-12 medium-6 columns`}>
        <PlotSettingsDropdown
          labelText={`t-SNE Perplexity`}
          options={perplexityOptions}
          defaultValue={{value: selectedPerplexity, label: selectedPerplexity}}
          onSelect={(selectedOption) => {
            onChangePerplexity(selectedOption.value)
          }}/>
      </div>
      }
      <div className={`small-12 medium-6 columns`}>
        <PlotSettingsDropdown
          labelText={`Colour plot by:`}
          options={metadata ? options : kOptions} // Some experiments don't have metadata in Solr, although they should do. Leaving this check in for now so we don't break the entire experiment page.
          defaultValue={defaultValue}
          onSelect={(selectedOption) => {
            onChangeColourBy(selectedOption.group, selectedOption.value)
          }}/>
      </div>
    </div>,
    <ScatterPlotLoader
      key={`cluster-plot`}
      wrapperClassName={`row`}
      chartClassName={`small-12 columns`}
      series={_colourizeClusters(highlightClusters)(plotData.series)}
      highchartsConfig={highchartsConfig}
      loading={loading}
      resourcesUrl={resourcesUrl}
      errorMessage={errorMessage}
    />
  ]
}

ClusterTSnePlot.propTypes = {
  height: PropTypes.number.isRequired,
  showControls: PropTypes.bool.isRequired,

  plotData: PropTypes.shape({
    series: PropTypes.array.isRequired
  }),
  highlightClusters: PropTypes.array,

  ks: PropTypes.arrayOf(PropTypes.number).isRequired,
  metadata: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string
  })),
  selectedColourBy: PropTypes.string,
  onChangeColourBy: PropTypes.func,
  clusterType: PropTypes.string,

  perplexities: PropTypes.arrayOf(PropTypes.number).isRequired,
  selectedPerplexity: PropTypes.number.isRequired,
  onChangePerplexity: PropTypes.func.isRequired,

  loading: PropTypes.bool.isRequired,
  resourcesUrl: PropTypes.string,
  errorMessage: PropTypes.string,

  tooltipContent: PropTypes.func
}

export {ClusterTSnePlot as default, _colourizeClusters, tooltipHeader}
