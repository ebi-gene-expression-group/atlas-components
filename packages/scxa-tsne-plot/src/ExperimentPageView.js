import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'
import queryString from 'query-string'

import {_colourizeClusters, _colourizeExpressionLevel} from './SeriesMapper'
import TSnePlot from 'expression-atlas-tsne-plot'
import AtlasAutocomplete from 'atlas-autocomplete'

class ExperimentPageView extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      params: queryString.parse(props.location.search),
      geneId: ``,
      k: 0,
      clusterId: [],
    }
  }

  handleChange(param, item) {
    let _newparam = {}
    _newparam[param] = (param === `k` ? Number.parseInt(item.target.value) : item)
    this.setState(_newparam)

    this.props.history.push(
      `?` +
      queryString.stringify({
        geneId: (param === `geneId` ? item : this.state.geneId),
        k: (param === `k` ? item.target.value : this.state.k)})
      + `&clusterId=[${this.state.clusterId}]`
    )
  }

  componentDidMount() {
    this.setState({
      geneId: this.state.params.geneId,
      k: this.state.params.k !== undefined ? Number.parseInt(this.state.params.k) : this.state.k,
      clusterId: JSON.parse(this.state.params.clusterId || `[]`)
    })
  }

  render() {
    const clusterOptions = this.props.availableClusters.sort()
      .map((name, ix) => (
        <option key={ix} value={name}>{name}</option>
      ))

    const selectedCluster = this.state.k ? this.state.k : this.props.availableClusters[0]

    const clusterHighchartsConfig = {
      plotOptions: {
        scatter: {
          tooltip: {
            headerFormat: `<b>Cluster {series.name}</b><br>`,
            pointFormat: `{point.name}`
          }
        }
      }
    }

    const geneExpressionHighchartsConfig = {
      plotOptions: {
        scatter: {
          tooltip: {
            headerFormat: `<b>{point.key} â€“ Cluster {series.name}</b><br>`,
            pointFormat: `Expression level: {point.expressionLevel}`
          }
        }
      }
    }

    return (
      <div className={`row`}>

        <div className={`small-6 columns`}>
          <label>Clustering: {this.state.k}</label>
          <select value={selectedCluster} onChange={this.handleChange.bind(this, `k`)}>
            {clusterOptions}
          </select>
          <TSnePlot atlasUrl={this.props.atlasUrl}
                    sourceUrl={`json/experiments/${this.props.experimentAccession}/tsneplot/clusters/${this.state.k}`}
                    seriesMapper={_colourizeClusters(this.props.highlightClusters)}
                    resourcesUrl={this.props.resourcesUrl}
                    highchartsConfig={clusterHighchartsConfig}/>
        </div>

        <div className={`small-6 columns`}>
          <AtlasAutocomplete atlasUrl={this.props.atlasUrl}
                             suggesterEndpoint={this.props.suggesterEndpoint}
                             enableSpeciesFilter={false}
                             initialValue={this.state.geneId}
                             onSelect={this.handleChange.bind(this, `geneId`)}/>
          <TSnePlot atlasUrl={this.props.atlasUrl}
                    sourceUrl={`json/experiments/${this.props.experimentAccession}/tsneplot/expression/${this.state.geneId}`}
                    seriesMapper={_colourizeExpressionLevel(this.props.expressionColourHue)}
                    resourcesUrl={this.props.resourcesUrl}
                    highchartsConfig={geneExpressionHighchartsConfig}/>
        </div>

      </div>
    )
  }
}

ExperimentPageView.propTypes = {
  props: PropTypes.object,
  atlasUrl: PropTypes.string,
  resourcesUrl: PropTypes.string,
  suggesterEndpoint: PropTypes.string,
  availableClusters: PropTypes.array.isRequired,
  highlightClusters: PropTypes.array,
  expressionColourHue: PropTypes.number
}

export default ExperimentPageView
