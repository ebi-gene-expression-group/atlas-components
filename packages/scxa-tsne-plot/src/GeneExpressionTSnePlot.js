/* eslint-disable react/no-children-prop */
import React from 'react'
import PropTypes from 'prop-types'
import Color from 'color'

import ScatterPlotLoader from './plotloader/PlotLoader'
import AtlasAutocomplete from 'expression-atlas-autocomplete'

import './util/MathRound'
import Responsive from 'react-responsive'


const Desktop = props => <Responsive {...props} minWidth={1800} />
const Tablet = props => <Responsive {...props} minWidth={1000} maxWidth={1799} />
const Mobile = props => <Responsive {...props} minWidth={767} maxWidth={999} />
const Default = props => <Responsive {...props} maxWidth={766} />


const _colourize = (colourRanges, defaultColour = `blue`, alpha = 0.65) => {
  return (val) => {
    if (isNaN(val)) {
      return Color(defaultColour).alpha(alpha).rgb().toString()
    }

    if (val === 0) {
      return Color(`lightgrey`).alpha(alpha).rgb().toString()
    }

    if (val > 9999) {
      return Color(`rgb(0, 0, 115)`).alpha(alpha).rgb().toString()
    }

    const rangeIndex = val <= 0 ? 0 : colourRanges.findIndex((colourRange) => colourRange.threshold >= val) - 1

    const loColour = Color(colourRanges[rangeIndex].colour)
    const hiColour = Color(colourRanges[rangeIndex + 1].colour)

    const redDelta = hiColour.red() - loColour.red()
    const greenDelta = hiColour.green() - loColour.green()
    const blueDelta = hiColour.blue() - loColour.blue()
    const increment = (val - colourRanges[rangeIndex].threshold) / (colourRanges[rangeIndex + 1].threshold - colourRanges[rangeIndex].threshold)

    return Color(
      `rgb(` +
      `${Math.floor(loColour.red() + redDelta * increment)}, ` +
      `${Math.floor(loColour.green() + greenDelta * increment)}, ` +
      `${Math.floor(loColour.blue() + blueDelta * increment)})`
    ).alpha(alpha).rgb().toString()
  }
}


const _colourizeExpressionLevel = (gradientColours, highlightSeries) => {
  const colourize = _colourize(gradientColours)
  return (plotData) => plotData.series.map((aSeries) => {
    // I canâ€™t think of a better way to reconcile series.name being a string and highlightSeries being an array of
    // numbers. For more flexibility we might think of having our series be identified by an arbitrary ID string

    if (!highlightSeries.length || highlightSeries.map((hs) => String(hs)).includes(aSeries.name)) {
      return {
        name: aSeries.name,
        data: aSeries.data.map((point) => {
          if(point.expressionLevel > 0){
            return {
              ...point,
              expressionLevel: Math.round10(point.expressionLevel, -2),
              colorValue: Math.round10(point.expressionLevel, -2)
            }
          } else {
            return {
              ...point,
              expressionLevel: Math.round10(point.expressionLevel, -2),
              color: Color(`lightgrey`).alpha(0.65).rgb().toString()
            }
          }

        })
      }
    } else {
      return {
        name: aSeries.name,
        data: aSeries.data.map((point) => ({
          ...point,
          expressionLevel: Math.round10(point.expressionLevel, -2),
          color: Color(`lightgrey`).alpha(0.65).rgb().toString()
        }))
      }
    }
  })
}

const GeneExpressionScatterPlot = (props) => {
  const {atlasUrl, suggesterEndpoint, geneId, onSelectGeneId, speciesName} = props       // Suggester
  const {height, plotData, expressionGradientColours, highlightClusters} = props  // Chart
  const {loading, resourcesUrl, errorMessage, showControls} = props                       // Overlay
  const colourSchema = [`#d4e4fb`,`#95adde`,`#6077bf`,`#1151D1`,`#35419b`,`#0e0573`] // light blue to dark blue
  const colourSchemaLength = colourSchema.length
  const plotIsDisabled = !plotData.max

  const dataScale = plotIsDisabled ?
    0 :
    plotData.max.toFixed(0).toString().length // Number of digits before decimal point
  const highchartsConfig = {
    chart: {
      // Magic number to adjust the height discrepancy between the two charts
      height: plotIsDisabled ? height - 12 : height
    },
    title: {
      text: `Gene expression`
    },
    tooltip: {
      formatter: function(tooltip) {
        // Trick Highcharts into thinking the point is in the bottom half of the chart, so that the tooltip
        // is displayed below the point
        this.point.negative = true

        const header = `<b>Cell ID:</b> ${this.point.name}<br>`
        const text = geneId ?
          `<b>Expression level:</b> ${this.point.expressionLevel} ${plotData.unit}` :
          `No gene selected`

        return header + text
      }
    },
    marker: {
      symbol: `circle`
    },
    colorAxis: plotIsDisabled ?
      {} :
      {
        min: 0.1,
        max: 10 ** dataScale - 1,
        type: `logarithmic`,
        reversed: false,
        //Dynamic stop where the last change colour is 100K
        stop: colourSchema.map((val,idx) => {
          return idx <= (Math.min(dataScale,colourSchemaLength) - 1) ?
            [(idx + 1)/Math.min(dataScale,colourSchemaLength),val] : []
        }),
        minColor:`rgb(215, 255, 255)`,
        maxColor: dataScale > colourSchemaLength ?
          colourSchema[colourSchemaLength - 1] : colourSchema[dataScale - 1],
        marker: {
          color: `#c4463a`
        }
      },
    legend: plotIsDisabled ?
      {
        enabled: false
      } :
      {
        title: {
          text: `Expression level (TPM)`
        },
        floating: false,
        align: `center`,
        symbolHeight: 5,
        symbolWidth: 450
      }
  }

  const responsiveComponent = width =>
    <ScatterPlotLoader
      key={`expression-plot`}
      wrapperClassName={`row`}
      chartClassName={`small-12 columns`}
      series={_colourizeExpressionLevel(expressionGradientColours, highlightClusters)(plotData)}
      highchartsConfig={highchartsConfig}
      loading={loading}
      legendWidth={width}
      resourcesUrl={resourcesUrl}
      errorMessage={errorMessage}
    />

  return [
    showControls &&
      <AtlasAutocomplete
        key={`expression-autocomplete`}
        wrapperClassName={`row margin-bottom-large`}
        atlasUrl={atlasUrl}
        suggesterEndpoint={suggesterEndpoint}
        initialValue={geneId}
        defaultSpecies={speciesName}
        onSelect={(event) => {
          onSelectGeneId(event)
        }}
      />,
    //react-responsive design
    <div key={`scatter-plot`} style={ showControls ? {} : {marginTop: `13%`}}>
      <Desktop key={`Desktop`}>
        {responsiveComponent(1800 / 2.2)}
      </Desktop>
      <Tablet key={`Tablet`}>
        {responsiveComponent(1000 / 2.2)}
      </Tablet>
      <Mobile key={`Mobile`}>
        {responsiveComponent(750)}
      </Mobile>
      <Default key={`Default`}>
        {responsiveComponent(350)}
      </Default>
    </div>
  ]
}

GeneExpressionScatterPlot.propTypes = {
  height: PropTypes.number.isRequired,
  showControls: PropTypes.bool.isRequired,

  plotData: PropTypes.shape({
    series: PropTypes.array.isRequired,
    unit: PropTypes.string.isRequired,
    max: PropTypes.number,
    min: PropTypes.number
  }),
  expressionGradientColours: PropTypes.arrayOf(PropTypes.shape({
    colour: PropTypes.string.isRequired,
    threshold: PropTypes.number.isRequired,
    stopPosition: PropTypes.number.isRequired
  })).isRequired,
  highlightClusters: PropTypes.array,

  atlasUrl: PropTypes.string.isRequired,
  suggesterEndpoint: PropTypes.string.isRequired,
  speciesName: PropTypes.string.isRequired,
  onSelectGeneId: PropTypes.func.isRequired,

  loading: PropTypes.bool.isRequired,
  resourcesUrl: PropTypes.string,
  errorMessage: PropTypes.string
}

GeneExpressionScatterPlot.defaultProps = {
  expressionGradientColours: [
    {
      colour: `rgb(215, 255, 255)`,
      threshold: 0,
      stopPosition: 0
    },
    {
      colour: `rgb(128, 255, 255)`,
      threshold: 10,
      stopPosition: 20
    },
    {
      colour: `rgb(0, 85, 225)`,
      threshold: 100,
      stopPosition: 40
    },
    {
      colour: `rgb(0, 0, 115)`,
      threshold: 10000,
      stopPosition: 100
    }
  ]
}

export {GeneExpressionScatterPlot as default, _colourizeExpressionLevel}
