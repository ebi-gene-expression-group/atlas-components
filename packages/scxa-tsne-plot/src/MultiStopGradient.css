.gradient {
    width: 20px;
    vertical-align: middle;
    margin: auto;
}

.tick {
    position: absolute;
    height: 2px;
    width: 20px;
}

.tickLabel {
    position: absolute;
    margin-top: -12px;
}

.left {
    composes: tickLabel;
    right: 20px;
    padding-right: 5px;
}

.right {
    composes: tickLabel;
    left: 20px;
    padding-left: 5px;
}