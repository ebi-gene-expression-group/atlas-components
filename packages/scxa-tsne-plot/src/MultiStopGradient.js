import React from 'react'
import PropTypes from 'prop-types'

import ScientificNotationNumber from 'expression-atlas-number-format'

import './util/MathRound'

// ranges have at least threshold (in ascending order) and stopPosition [0..100]
const putInRange = (ranges, val) => {
  if (val <= ranges[0].threshold) {
    return 0
  }

  if (val >= ranges[ranges.length - 1].threshold) {
    return 1
  }

  const rangeIndex = ranges.findIndex((range) => range.threshold >= val) - 1

  const withinRangeOffset =
    (val - ranges[rangeIndex].threshold) / (ranges[rangeIndex + 1].threshold - ranges[rangeIndex].threshold)

  return ranges[rangeIndex].stopPosition * (1 + withinRangeOffset) / 100
}



const MultiStopGradient = ({height, colourRanges, plotData}) => {
  const bg = colourRanges.map((colourRange) => `${colourRange.colour} ${colourRange.stopPosition}%`).join(`, `)
  const minMaxExpressionTickHeight = 2
  const gradientHeight = height - 100

  const minExpressionBottom =
    Math.min(
      gradientHeight - minMaxExpressionTickHeight,
      gradientHeight - gradientHeight * putInRange(colourRanges, plotData.min))
  const maxExpressionBottom =
    Math.max(
      minMaxExpressionTickHeight,
      gradientHeight - gradientHeight * putInRange(colourRanges, plotData.max))

  return (
    <div className={`small-2 columns text-center`}>
      <div>
        <small><ScientificNotationNumber value={colourRanges[colourRanges.length - 1].threshold}/> {plotData.unit}</small>
      </div>

      <div
        style={{
        width: `20px`,
        height: `${gradientHeight}px`,
        background: `linear-gradient(0deg, ${bg})`,
        verticalAlign: `middle`,
        margin: `auto`}}>

        <div style={{position: `relative`, height: `2px`, width: `20px`, background: `grey`, top: `${maxExpressionBottom}px`}}>
          <div style={{position: `absolute`, marginLeft: `24px`, marginTop: `-0.75rem`}}>
            <small><ScientificNotationNumber value={Math.round10(plotData.max, -2)}/></small>
          </div>
        </div>
        <div style={{position: `relative`, height: `2px`, width: `20px`, background: `grey`, top: `${minExpressionBottom}px`}}>
          <div style={{position: `absolute`, marginLeft: `24px`, marginTop: `-0.75rem`}}>
            <small><ScientificNotationNumber value={Math.round10(plotData.min, -2)}/></small>
          </div>
        </div>

      </div>

      <div>
        <small><ScientificNotationNumber value={colourRanges[0].threshold}/> {plotData.unit}</small>
      </div>
    </div>
  )
}

MultiStopGradient.propTypes = {
  height: PropTypes.number.isRequired,
  colourRanges: PropTypes.arrayOf(PropTypes.shape({
    colour: PropTypes.string.isRequired,
    threshold: PropTypes.number.isRequired
  })).isRequired,
  plotData: PropTypes.shape({
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    unit: PropTypes.string.isRequired
  })
}

export default MultiStopGradient
