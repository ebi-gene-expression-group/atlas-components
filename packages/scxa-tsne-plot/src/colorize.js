import Color from 'color'

// const defaultColor = `#eeeeee`
// const stops = [
//   // The first stop acts as a threshold: anything lower than this gets the defaultColor
//   {
//     cutoff: 0.1,
//     color: `#e1e9f4`
//   },
//   {
//     cutoff: 1,
//     color: `#d4e4fb`
//   },
//   {
//     cutoff: 10,
//     color: `#95adde`
//   },
//   {
//     cutoff: 100,
//     color: `#6077bf`
//   },
//   {
//     cutoff: 1000,
//     color: `#1151d1`
//   },
//   {
//     cutoff: 10000,
//     color: `#35419b`
//   },
//   {
//     cutoff: 100000,
//     color: `#0e0573`
//   },
//   {
//     cutoff: 1000000,
//     color: `#07004c`
//   }
// ]

const colorize = (stops, defaultColour) =>
  (val) => {
    // Below minimum
    if (val < stops[0].cutoff) {
      return Color(defaultColour).alpha(0.65).toString()
    }
    // Above maximum
    if (val > stops[stops.length - 1].cutoff) {
      return Color(stops[stops.length - 1].color).toString()
    }

    // General case
    const stopIndex = stops.findIndex(stop => val <= stop.cutoff)

    const minColor = Color(stops[Math.max(0, stopIndex - 1)].color)  // Because stopIndex may be 0
    const maxColor = Color(stops[stopIndex].color)

    // Transform colors to RGB
    const positionWithinStops = val / stops[stopIndex].cutoff
    return Color.rgb(
      minColor.red() + (maxColor.red() - minColor.red()) * positionWithinStops,
      minColor.green() + (maxColor.green() - minColor.green()) * positionWithinStops,
      minColor.blue() + (maxColor.blue() - minColor.blue()) * positionWithinStops,
    ).toString()
  }

export default colorize
