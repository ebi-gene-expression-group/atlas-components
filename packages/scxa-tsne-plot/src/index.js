import ExperimentPageView from './ExperimentPageView'

export default ExperimentPageView
