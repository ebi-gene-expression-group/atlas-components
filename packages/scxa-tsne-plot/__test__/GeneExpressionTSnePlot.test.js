import React from 'react'
import renderer from 'react-test-renderer'
import Color from 'color'

import Enzyme from 'enzyme'
import {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import {_colourizeExpressionLevel} from '../src/GeneExpressionTSnePlot'
import GeneExpressionTSnePlot from '../src/GeneExpressionTSnePlot'
import MultiStopGradient from '../src/MultiStopGradient'

import '../src/util/MathRound'
import {gradientColourRanges, randomHighchartsSeries, randomHighchartsSeriesWithNamesAndMaxPoints, plotData, randomHighchartsSeriesWithSeed} from './Utils'

Enzyme.configure({ adapter: new Adapter() })

describe(`GeneExpressionTSnePlot colourize function`, () => {

  test(`must not change the number of series`, () => {
    const seriesNames = [`0`, `1`, `2`, `3`, `4`]
    const maxPointsPerSeries = 1000
    const randomSeries = randomHighchartsSeriesWithNamesAndMaxPoints(seriesNames, maxPointsPerSeries)
    expect(_colourizeExpressionLevel(gradientColourRanges(), [])(plotData(randomSeries))).toHaveLength(seriesNames.length)
  })

  test(`must not change the number of points in each series`, () => {
    const randomSeries = randomHighchartsSeries()
    _colourizeExpressionLevel(gradientColourRanges(), [])(plotData(randomSeries)).forEach((series, i) => {
      expect(series.data).toHaveLength(randomSeries[i].data.length)
    })
  })

  test(`adds a color field to all points`, () => {
    const randomSeries = randomHighchartsSeries()

    _colourizeExpressionLevel(gradientColourRanges(), [])(plotData(randomSeries)).forEach((series) => {
      series.data.forEach((point) => {
        expect(point).toHaveProperty(`color`)
      })
    })
  })

  test(`assigns maximum colour to the point with highest expression`, () => {
    const randomSeries = randomHighchartsSeries()
    randomSeries[randomSeries.length - 1].data.push({
      x: 0,
      y: 0,
      expressionLevel: 10000,
      name: "Maximum overkill"
    })

    const allPoints = randomSeries.reduce((acc, series) => acc.concat(series.data), [])
    const maxExpressionLevel = Math.round10(Math.max(...allPoints.map((point) => point.expressionLevel)), -2)

    const maxExpressionLevelPoints = _colourizeExpressionLevel(gradientColourRanges(), [])(plotData(randomSeries)).reduce((acc, series) => {
      acc.push(series.data.filter((point) => point.expressionLevel === maxExpressionLevel, -2))
      return acc
    }, [])
    .reduce((acc, points) => points.length ? acc.concat(points) : acc, [])

    expect(maxExpressionLevelPoints.length).toBeGreaterThanOrEqual(1)
    maxExpressionLevelPoints.forEach((point) => {
      expect(point).toHaveProperty(`color`, Color(gradientColourRanges()[gradientColourRanges().length - 1].colour).alpha(0.65).rgb().toString())
    })
  })

  test(`assigns grey colour to the point with 0 expression`, () => {
    const randomSeries = randomHighchartsSeries()
    randomSeries[randomSeries.length - 1].data.push({
      x: 0,
      y: 0,
      expressionLevel: 0,
      name: "Minimum underkill"
    })

    const allPoints = randomSeries.reduce((acc, series) => acc.concat(series.data), [])
    const minExpressionLevel = Math.round10(Math.min(...allPoints.map((point) => point.expressionLevel)), -2)

    const minExpressionLevelPoints = _colourizeExpressionLevel(gradientColourRanges(), [])(plotData(randomSeries)).reduce((acc, series) => {
      acc.push(series.data.filter((point) => point.expressionLevel === minExpressionLevel, -2))
      return acc
    }, [])
    .reduce((acc, points) => points.length ? acc.concat(points) : acc, [])

    expect(minExpressionLevelPoints.length).toBeGreaterThanOrEqual(1)
    minExpressionLevelPoints.forEach((point) => {
      expect(point).toHaveProperty(`color`, Color('lightgrey').alpha(0.65).rgb().toString())
    })
  })

  test(`rounds expression level to two decimal places`, () => {
    const randomSeries = randomHighchartsSeries()

    _colourizeExpressionLevel(gradientColourRanges(), [])(plotData(randomSeries)).forEach((series) => {
      series.data.forEach((point) => {
          if (String(point.expressionLevel).includes(`.`)) {
            expect(String(point.expressionLevel).split(`.`)[1].length).toBeLessThanOrEqual(2)
          }
        })
    })
  })

  test(`assigns default colour, i.e. blue, if points have no expression level property`, () => {
    _colourizeExpressionLevel(gradientColourRanges(), [])({
      series: [
        {
          name: `Cluster 1`,
          data: [
            {
              name: `Point 1-1`,
              x: 1,
              y: 1
            }
          ]
        },
        {
          name: `Cluster 2`,
          data: [
            {
              name: `Point 2-1`,
              x: 2,
              y: 2
            },
            {
              name: `Point 2-2`,
              x: 3,
              y: 3
            }
          ]
        }
      ],
      min: 100.0,
      max: 100.0
    }).forEach((series) => {
      series.data.forEach((point) => {
          expect(point).toHaveProperty(`color`, Color(`blue`).alpha(0.65).rgb().toString())
        })
    })
  })
})

describe(`GeneExpressionTSnePlot`, () => {
  test(`with random data matches snapshot`, () => {
    const randomSeries = randomHighchartsSeriesWithSeed()
    const onSelectGeneId = () => {}

    const tree = renderer
      .create(<GeneExpressionTSnePlot height={600} expressionGradientColours={gradientColourRanges()} atlasUrl={``} suggesterEndpoint={``} onSelectGeneId={onSelectGeneId} loading={true} plotData={plotData(randomSeries)} highlightClusters={[]}/>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test(`contains MultiStopGradient component when min and max values exist`, () => {
    const randomSeries = randomHighchartsSeriesWithSeed()
    const onSelectGeneId = () => {}

    const wrapper = mount(<GeneExpressionTSnePlot height={600} expressionGradientColours={gradientColourRanges()} atlasUrl={``} suggesterEndpoint={``} onSelectGeneId={onSelectGeneId} loading={true} plotData={plotData(randomSeries)} highlightClusters={[]}/>)

    expect(wrapper.find(MultiStopGradient).length).toBe(1)
  })
})
