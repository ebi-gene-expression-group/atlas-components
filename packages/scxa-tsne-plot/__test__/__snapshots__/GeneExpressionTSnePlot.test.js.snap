// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneExpressionTSnePlot with random data matches snapshot 1`] = `
Array [
  <div
    className="row margin-bottom-large"
  >
    <label>
      Gene ID, gene name or gene feature
    </label>
    <div
      style={
        Object {
          "display": "",
        }
      }
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        name="geneId"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        type="text"
        value=""
      />
    </div>
  </div>,
  <div
    style={Object {}}
  />,
]
`;
